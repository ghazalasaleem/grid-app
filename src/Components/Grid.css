.table-container{
    border: 2px solid #edf3f5;
    padding: 5px;
    margin: 10px;
    box-shadow: 2px 2px 3px gainsboro;
    color: #5c6475;
}
.table-header{
    background-color: #fcfcfc;
    /* text-align: center; */
}
.grid-border{
    border: 1px solid #edf3f5;
}
.table-row{
    width: 100%;
    display: table;
    table-layout: fixed;
}
.table-cell{
    display: table-cell;
    border: 1px solid #edf3f5;
    padding: 3px 2px 3px 3px;
}
.table-cell > input,  .table-cell > select{
    width: 94%;
    border: 1px solid white;
}
input, select {
    border-radius: 4px;
    border: 1px solid #edf3f5;
    color: #5c6475;
}
input:focus, select:focus{
    border: 1px solid darkgray;
    border-radius: 4px;
    outline: none;
}
.large-table{
width: 50%;
}
.pad-l5{
    padding-left: 5px;
}
.cursor-pointer{
    cursor: pointer;
}
.vertical-middle{
    display: inline;
    vertical-align: middle;
}
.pad-7{
    padding: 7px;
}
.pad-5{
    padding: 5px;
}
.mar-r10{
    margin-right: 10px;
}
label{
    margin-right: 7px;
}
div[data-title="action"]{
    width: 15%;
}

.bootstrap-select .bs-ok-default::after {
    width: 0.3em;
    height: 0.6em;
    border-width: 0 0.1em 0.1em 0;
    transform: rotate(45deg) translateY(0.5rem);
}

.btn.dropdown-toggle:focus {
    outline: none !important;
}