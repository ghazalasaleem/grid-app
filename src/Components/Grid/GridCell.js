import React from 'react';
const GridCell = props =>{

    const {cellData, data} = props;
    const {key, cell} = cellData;
    const {id} = data;
    const value = data[key];
    let cellDOM;
    if(cell && typeof cell.renderer === "function"){
        cellDOM= cell.renderer({data: data, key: key});
    }
    else{
        cellDOM=<label>{value}</label>;
    }
        return (<div className="table-cell" key={id} data-id={id} data-key={key}>{cellDOM}</div>);
};

export default GridCell;