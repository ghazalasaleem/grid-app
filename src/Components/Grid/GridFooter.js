import React, { useContext } from 'react';
import Pagination from 'react-js-pagination';
import GridContext from './Context/GridContext';

const GridFooter = props =>{

    const gridCntxt = useContext(GridContext);
    const {activePage, rowsPerPage, pageRange, totalRows, onPageChange}= gridCntxt;

    return (
    <div className="table-footer table-row">
        <div className="table-cell">
            <Pagination
            activePage={activePage}
            itemsCountPerPage={rowsPerPage}
            totalItemsCount={totalRows}
            pageRangeDisplayed={pageRange}
            onChange={(e)=>{onPageChange(e)}}
            itemClass="page-item"
            linkClass="page-link">
            </Pagination>
        </div>
    </div>);
}

export default GridFooter;