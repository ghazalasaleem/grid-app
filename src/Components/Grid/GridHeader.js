import React, { useContext } from 'react';
import {FiArrowDown, FiArrowUp} from 'react-icons/fi';
import HeaderContext from './Context/HeaderContext';

const GridHeader = props =>{

    const { handleSort, sortCol, sortOrderAsc} = props;
    const headerList = useContext(HeaderContext)
    
    return (
        headerList.map(header =>{
            const classes = "table-cell" + (header.sort && header.sort.active?" cursor-pointer":" events-none");
            return header.show?
                <div className={classes} key={header.key} data-title={header.key} onClick={e=>handleSort(e, header.key)}>
                    {header.name}
                    {(header.sort && header.sort.active)?
                        <div className="pad-l5 vertical-middle">
                            {(sortCol === header.key && !sortOrderAsc)?<FiArrowUp/>:<FiArrowDown/>}
                        </div>:''
                    }
                </div>:'';
        })
    );
};

export default GridHeader;