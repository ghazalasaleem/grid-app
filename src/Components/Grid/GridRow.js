import React, {useContext } from 'react';
import GridCell from './GridCell';
import GridContext from './Context/GridContext';

 const GridRow = props =>{

    const {data, renderList} = props;
    const gridCntxt = useContext(GridContext);
    const {rowSelection, handleRowSelection} = gridCntxt;

    return (
        !data.hidden && (
        <div className={data.selected?"table-row selected":"table-row"} key={data.id} data-id={data.id}>
            {rowSelection && 
                (<div data-id={data.id} className="table-cell chkCol">
                    <input type="checkbox" checked={data.selected} onChange={e=>handleRowSelection(e,data.id)}/>
                </div> )}
                                      
            {renderList && renderList.length && 
                renderList.map(renderer =>{
                    return <GridCell key={data.id+renderer.key} cellData={renderer} data={data}></GridCell>;
                })
            }
        </div>
        )
    );
 }
 export default GridRow;