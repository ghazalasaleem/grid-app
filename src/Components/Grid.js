import React, { useEffect, useState } from 'react';
import './Grid.css'
import GridRow from './GridRow';
import GridToolbar from './GridToolbar';
import {FiArrowDown, FiArrowUp} from 'react-icons/fi';
import AppModal from './AppModal';

const Grid = props => {

    const {headers, tableData, newRow} = props;
    const [contentHeader, setContentHeader] = useState(headers);
    const [content, setContent] = useState([]);
    const [sortCol, setSortCol] = useState("");
    const [sortOrderAsc, setSortOrderAsc] = useState(true);
    const [modalData, setModalData] = useState(null)
    const [isInfoModalVisible, setIsInfoModalVisible] = useState(false);
    const[customCols, setCustomCols] = useState([]);

    useEffect(() =>{
        
        const sortableCol = headers.find(prop => prop.sort);
        const colName = sortableCol?sortableCol.name.toLowerCase():"";
        setSortCol(colName);
        setSortOrderAsc(true);
        
        setContent([...sortContentList(tableData, colName)]);
        setCustomCols(Array.from(headers,obj=> obj.name.toLowerCase()));
    },[tableData, headers]);

    useEffect(() =>{
        const data = [...tableData];
        const headerList = [...headers];
        let customData = [];
        data.map(item=>{
            let propList =[];
            item.map((prop)=>{
                if(customCols.find(col => col === prop.name)){
                    propList.push(Object.assign({visible: true},{...prop}));
                }
                else{
                    propList.push(Object.assign({visible: false},{...prop}));
                }
            });
            customData.push([...propList]);
        });
        setContent([...sortContentList(customData, sortCol)]);

        headerList.map(item =>{
            if(customCols.find(col => col === item.name.toLowerCase())){
                item.visible = true;
            }
            else{
                item.visible = false;
            }
        });

        setContentHeader(headerList);
    },[customCols.length]);

    const sortContentList = (list, key) =>{
        return key?(list.sort((a,b) => {
            let aVal = a.find(prop => prop.name === key).value.toLowerCase();
            let bVal = b.find(prop => prop.name === key).value.toLowerCase();
            if(aVal < bVal) return -1;
            else if(aVal > bVal) return 1;
            else return 0;

        })):[];
    };

    const handleSort = (event, key) =>{
        const sortKey = key? key.toLowerCase():'';
        let  sortOrder = sortOrderAsc;
        let tableContent =[...content];
        if(sortKey){
            if(sortKey !== sortCol) {
                setSortCol(sortKey);
                sortOrder = true;
                setSortOrderAsc(sortOrder);                
            } 
            else{
                sortOrder = !sortOrderAsc;
                setSortOrderAsc(sortOrder);
            } 
            tableContent = sortContentList(tableContent, sortKey);
            if(!sortOrder){
                tableContent.reverse();
            }
            setContent([...tableContent]);      
        }
    };
    const handleRowDelete = (e) =>{
        const index = e.currentTarget?e.currentTarget.getAttribute('data-id'):null;
        if(index >= 0){
            let tableContent = [...content];
            tableContent.splice(index, 1);
            setContent([...tableContent]); 
        }               
    };
    const handleRowCopy = (e) =>{
        let index = e.currentTarget?e.currentTarget.getAttribute('data-id'):null;
        if(index && parseInt(index) >= 0){
            index = parseInt(index);
            let copyRow =JSON.parse(JSON.stringify([...content][index]));
            copyRow.map((prop) =>{
                if(prop.name === "name") {
                    prop.key = "CC"+prop.key;
                    prop.value = "Copy of "+prop.value;
                }
            })
            let tableContent = [...content];
            tableContent.splice(++index,0,copyRow);
            setContent([...tableContent]);
        }
    };
    const handleRowInfo = (e) =>{
        let index = e.currentTarget?e.currentTarget.getAttribute('data-id'):null;
        if(index && parseInt(index) >= 0) {
            index = parseInt(index);
            let rowName =[...content][index].find(prop => prop.name === "name").value;
            setIsInfoModalVisible(true);
            setModalData(rowName);
        }
    };
    const handleAddRow = e =>{
        let dataList = [...content];
        dataList.splice(0,0,newRow);
        setContent([...sortContentList(dataList, sortCol)]);
    };
    const handleShowCols = e=>{
        const options = e.currentTarget.options;
        let value = [];
        for (var i = 0, l = options.length; i < l; i++) {
            if (options[i].selected) {
                value.push(options[i].value);
            }
        }
        setCustomCols(value);
    }
    return(
        <React.Fragment>
    <div className="large-table table-container">
        
        {headers && headers.length && (
            <React.Fragment>
            <GridToolbar showAddnew={true} showSort={false} showHide={true} headers={headers} handleHideColmns={handleShowCols} selectedList={customCols} handleAdd={handleAddRow}></GridToolbar>
            <div className="table-header table-row">
                {contentHeader.map(header =>{
                    const rowClass = "table-cell"+ (header.sort?" cursor-pointer":"");
                    const cellDom = header.visible !== false?
                    (<div className={rowClass} key={header.key} data-title={header.name.toLowerCase()} onClick={e => handleSort(e,header.name)}>{header.name}
                    {header.sort?<div className="pad-l5 vertical-middle">{sortOrderAsc?<FiArrowDown/>:<FiArrowUp/>}</div>:''}</div>
                ):"";
                    return cellDom;
                    
                    })}
            </div>
            </React.Fragment>
        )}
        {content && content.length && (
            <div className="table-body">
                {content.map((data,index) =>{
                    return (<GridRow 
                        data={data}
                        key={index}
                        index={index}
                        handleRowDelete={handleRowDelete}
                        handleRowCopy={handleRowCopy}
                        handleRowInfo={handleRowInfo}>

                        </GridRow>);
                })
                }
            </div>
            
        )}
    </div>
    {modalData && isInfoModalVisible && (
     <AppModal data={modalData}></AppModal>   
    )}
    </React.Fragment>
    );
}

export default Grid;