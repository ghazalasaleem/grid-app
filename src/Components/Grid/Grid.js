import React, { useEffect, useState } from 'react';
import './Grid.css'
import GridRow from './GridRow';
import GridHeader from './GridHeader';
import {GridProvider} from './Context/GridContext';
import GridFooter from './GridFooter';

const Grid = props => {

    const {configData, dataList, selectCallback, sortRowCallback, paginationCallback} = props;
    const [renderList, setRenderList] = useState([]);
    const [headerList, setHeaderList] = useState([]);
    const [content, setContent] = useState([]);
    const [sortCol, setSortCol] = useState("");
    const [sortOrderAsc, setSortOrderAsc] = useState(true);
    const [rowSelection, setRowSelection] = useState(false);
    const [selectAll, setSelectAll] = useState(false);
    const [selectedRows, setSelectedRows]  = useState([]);
    const [activePage, setActivePage] = useState(1);
    const [defaultSort, setDefaultSort] = useState({});
 
    useEffect(()=>{
        let headerL = [], renderL = [];
        if(configData.columns){
            configData.columns.map((row)=>{
                if(!defaultSort || !defaultSort.col){
                    if(row.sort && row.sort.default && row.sort.active){
                        setDefaultSort({col: row.key, order: row.sort.order})  
                        setSortCol(row.key);
                        setSortOrderAsc(row.sort.order === 'asc'?true:false);
                    }
                }
                headerL.push({
                    name: row.name,
                    key: row.key,
                    sort: row.sort,
                    show: true
                });
                renderL.push({
                    key: row.key,
                    cell: row.cell
                });
            });
        }
    
        let dl = dataList && dataList.length?[...dataList]:[];
        dl.map(obj=>{obj.selected = false});
        setHeaderList([...headerL]);
        setRenderList(renderL); 
        setContent([...dl]);
        setRowSelection(configData.rowSelection);
    },[configData, dataList]);

    useEffect(()=>{
        let dataL = JSON.parse(JSON.stringify(content));
        dataL.map((data)=>{
            data.selected = selectAll;
        });
        setContent([...dataL]);
    },[selectAll]);

    useEffect(()=>{
        if(selectCallback) selectCallback(selectedRows);
    },[selectedRows]);

    useEffect(()=>{
        setSelectedRows(getSelectedRows());
    },[content]);

    useEffect(()=>{
        if(paginationCallback) paginationCallback(activePage);
    },[activePage]);

    const handleRowSelection = (e, id) =>{
        if(e && e.currentTarget){
            const selectAllFlag = e.currentTarget.checked;
            if(id){
                let dataL = JSON.parse(JSON.stringify(content));
                dataL.map((data)=>{
                    if(data.id === id) data.selected = selectAllFlag;
                });
            setContent([...dataL]);
            }
            else{
                setSelectAll(selectAllFlag);
            }
        }
    };

    const handleSort = (e, key) =>{
        const sortKey = key? key.toLowerCase():'';
        let  sortOrder = sortOrderAsc;
        if(sortKey){
            if(sortKey !== sortCol) {
                setSortCol(sortKey);
                sortOrder = true;
                setSortOrderAsc(sortOrder);                
            } 
            else{
                sortOrder = !sortOrderAsc;
                setSortOrderAsc(sortOrder);
            } 
            sortRowCallback({key: sortKey, order: sortOrder?'asc':'desc'});      
        }
    };

    const pageChangeHandler = (pageNo) =>{
        // console.log('selected pg no -', pageNo);
        setActivePage(pageNo);
        setSortCol(defaultSort.col);
        setSortOrderAsc(defaultSort.order === 'asc'? true: false);
    }

    const getSelectedRows = () =>{
        let selectedList = [];
        content.map(data=>{
            if(data.selected) selectedList.push(JSON.parse(JSON.stringify(data)));
        });

        return selectedList;
    };

    return(
        <React.Fragment>
            <GridProvider value={{
                sortCol: sortCol,
                sortOrderAsc: sortOrderAsc,
                activePage: activePage,
                rowsPerPage: configData.rowsPerPage,
                totalRows: configData.totalRows,
                rowSelection: rowSelection,
                pageRange: configData.pageRange,
                handleSort:handleSort,
                handleRowSelection: handleRowSelection,
                onPageChange: pageChangeHandler
            }}>
            <div className="large-table table-container">
                {headerList && headerList.length && (
                    <div className="table-header table-row">
                        {
                            rowSelection && 
                            (<div className="table-cell chkCol">
                                <input type="checkbox" className="" checked={selectAll} onChange={(e)=>{handleRowSelection(e)}}/>
                            </div> )
                        }
                        <GridHeader headerList={headerList}></GridHeader>
                    </div>
                )}
                {content && content.length && (
                    <div className="table-body">
                        {content.map((data) =>{
                           return ( 
                                <GridRow data={data} key={data.id} renderList={renderList}></GridRow>                            
                           );
                        })}
                    </div>        
                )}
                {configData.pagination && (
                    <GridFooter></GridFooter>
                )}                
            </div>
            </GridProvider>
        </React.Fragment>
    );
}

export default Grid;