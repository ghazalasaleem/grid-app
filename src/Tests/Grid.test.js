import React from 'react';
import renderer from 'react-test-renderer';
import Grid from '../Components/Grid/Grid';
import {shallow, mount, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
describe('Grid Component', () =>{
  let grid;
  
    beforeAll(() => {
      grid = renderer.create(<Grid/>).toJSON();
    });
    afterAll(() => {
      grid = null;
    });
     
    test('Grid renders', () => {
      expect(grid).toBeDefined();
      expect(grid).toMatchSnapshot();
    });

    test('with props', () => {

    });
});

