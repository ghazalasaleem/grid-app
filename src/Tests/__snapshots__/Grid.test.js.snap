// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Component Grid renders 1`] = `
<div
  className="large-table table-container"
>
  0
  0
</div>
`;
